src/code/PreRender.c

```diff
@@ -254,150 +350,339 @@ void func_80170AE0(PreRender* this, Gfx** gfxp, s32 alpha) {
+                PreRender_AntiAliasFilter(this, x, y);
             }
         }
     }
 }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/PreRender/func_801716C4.s")
+/**
+ * Selects the median value from 9 different 5-bit pixels:
+ * px1[0], px1[1], px1[2], px2[0], px2[1], px2[2], px3[0], px3[1], px3[2]
+ * all args are expected to be an array of 3 different 5-bit values
+ */
+u32 PreRender_Get5bMedian9(u8* px1, u8* px2, u8* px3) {
+    u8 pxValCount[32]; // Stores the count for each of the possible 32 5-bit pixel values
+    u32 pxCount;       // Pixel count
+    s32 pxMed;         // Pixel median value
+
+    // Initialize count to 0 in groups of 4 bits
+    *(s32*)(&pxValCount[0]) = 0;
+    *(s32*)(&pxValCount[4]) = 0;
+    *(s32*)(&pxValCount[8]) = 0;
+    *(s32*)(&pxValCount[12]) = 0;
+    *(s32*)(&pxValCount[16]) = 0;
+    *(s32*)(&pxValCount[20]) = 0;
+    *(s32*)(&pxValCount[24]) = 0;
+    *(s32*)(&pxValCount[28]) = 0;
+
+    // Increment the count that contains the pixel values
+    pxValCount[px1[0]]++;
+    pxValCount[px1[1]]++;
+    pxValCount[px1[2]]++;
+
+    pxValCount[px2[0]]++;
+    pxValCount[px2[1]]++;
+    pxValCount[px2[2]]++;
+
+    pxValCount[px3[0]]++;
+    pxValCount[px3[1]]++;
+    pxValCount[px3[2]]++;
+
+    // Loop through the 32 bits until 5 values are found, then return that bit value.
+    // Note that the median of 9 is the 5th sequential value.
+    pxCount = 0;
+    pxMed = 0;
+    while (true) {
+        pxCount += pxValCount[pxMed];
+        if (pxCount >= 5) {
+            // the median is found
+            break;
+        }
+        pxMed++;
+    }
+
+    return pxMed;
+}
+
+void PreRender_DivotFilter(PreRender* this) {
+    u32 width = this->width;
+    u32 height = this->height;
+    u8* buffer = alloca(width * 10);
+    u8* redRow[3];
+    u8* greenRow[3];
+    u8* blueRow[3];
+    u8* cvgFull;
+    Color_RGBA16 inPx;
+    Color_RGBA16 outPx;
+    u32 x;
+    u32 y;
+    s32 pad;
+
+    redRow[0] = &buffer[width * 0];
+    redRow[1] = &buffer[width * 1];
+    redRow[2] = &buffer[width * 2];
+
+    greenRow[0] = &buffer[width * 3];
+    greenRow[1] = &buffer[width * 4];
+    greenRow[2] = &buffer[width * 5];
+
+    blueRow[0] = &buffer[width * 6];
+    blueRow[1] = &buffer[width * 7];
+    blueRow[2] = &buffer[width * 8];
+
+    cvgFull = &buffer[width * 9];
+
+    // Fill line buffers for first 2 rows
+    for (y = 0; y < 2; y++) {
+        for (x = 0; x < width; x++) {
+            inPx.rgba = this->fbufSave[x + y * this->width];
+
+            redRow[y][x] = inPx.r;
+            greenRow[y][x] = inPx.g;
+            blueRow[y][x] = inPx.b;
+        }
+    }
+
+    // For each row in the image, except first and last
+    for (y = 1; y < height - 1; y++) {
+        // Find start of pixels and coverage for current line (bug? this should probably be fetching the NEXT line, but
+        // really the divot filter only cares about individual lines so it's already wrong)
```

Comment:
This comment is written in a weird way. Is this actually a bug or not?

Reply:
I might defer this comment to @Thar0 ?

Reply:
It's a bug in the sense that it probably doesn't do what the writer intended (if I understand what they were trying to do), but what the writer intended is also wrong with respect to a hardware-accurate divot filter.